### ISSUE TITLE
BUG: Extrinsic/intrinsic swapped in documentation for `spatial.transform.Rotation.from_euler()` ?

### ISSUE THREAD

**Issue by whophil** [scipy/scipy#17464]
### Describe your issue.

The documentation for `spatial.transform.Rotation.from_euler()` states that a `seq` with uppercase letters represents an intrinsic rotation, while a `seq` with lowercase letters represents an extrinsic rotation. Is it possible that this is swapped? (Or that I have a fundamental misunderstanding of extrinsic vs intrinsic rotations?)

As far as I can tell, this concern was first raised [in this comment](https://github.com/scipy/scipy/issues/10959#issuecomment-859295711), which provided a simple test case.

A gist with a different check is provided here: https://gist.github.com/whophil/c809692ceb8d5b4a22b70edc103dd87c, but arrives at the same conclusion.

### Reproducing Code Example

```python
see gist
```

### Error message

```shell
n/a
```

### SciPy/NumPy/Python version information

1.9.3 1.23.5 sys.version_info(major=3, minor=11, micro=0, releaselevel='final', serial=0)

**Comment by Patol75** [comment]
I am wondering if the issue here could be the order in which the angles are provided to the `from_euler` call. The following piece of code does not raise and it seems to me that `(yaw * pitch * roll).as_matrix()` replicates the assembly of the `R` matrix in the example you provide.
```
import numpy as np
from scipy.spatial.transform import Rotation as R

assert np.allclose(
    R.from_euler("xyz", [15, 10, 5], degrees=True).as_matrix(),
    R.from_euler("ZYX", [5, 10, 15], degrees=True).as_matrix(),
)

roll = R.from_euler("X", 15, degrees=True)
pitch = R.from_euler("Y", 10, degrees=True)
yaw = R.from_euler("Z", 5, degrees=True)

assert np.allclose(
    (yaw * pitch * roll).as_matrix(),
    R.from_euler("ZYX", [5, 10, 15], degrees=True).as_matrix(),
)

```

**Comment by adigitoleo** [comment]
I agree with @Patol75, the first rotation matrix from
https://en.wikipedia.org/w/index.php?title=Rotation_matrix&oldid=1122062121#General_rotations
represents an intrinsic rotation around Z, then Y, then X.
That wikipedia section is confusing for several reasons[^1].
Instead, I suggest these slightly better resources:
- http://flothesof.github.io/intrinsic-extrinsic-rotations-euler-angles.html#TL;DR
- https://www.mauriciopoppe.com/notes/computer-graphics/transformation-matrices/rotation/euler-angles/
- https://en.wikipedia.org/wiki/Davenport_chained_rotations#Conversion_to_extrinsic_rotations
- https://math.stackexchange.com/questions/1137745/proof-of-the-extrinsic-to-intrinsic-rotation-transform/3314025#3314025

The summary is that you appear to have mixed up intrinsic and extrinsic rotations.
For **extrinsic** rotations, the order of composition is **reversed**.
For **intrinsic** rotations, the order of composition is **forward** (left-to-right).
From what I understand, this comes from a second reversal of the ordering needed to express the intrinsically-rotated vector in the static "world" coordinate frame.

The `from_euler` methods to take arguments which specify a rotation $R$
composed of elementary rotations $R_1(α), R_2(β), R_3(γ)$ for angles {α, β, γ}.
See the following:

```python
In [1]: import numpy as np

In [2]: from sympy import symbols, Matrix, sin, cos

In [3]: theta, alpha, beta, gamma = symbols("theta, alpha, beta, gamma")

In [4]: rot_z = Matrix([
   ...: [cos(theta), -sin(theta), 0],
   ...: [sin(theta), cos(theta), 0],
   ...: [0, 0, 1],
   ...: ])

In [5]: rot_y = Matrix([
   ...: [cos(theta), 0, sin(theta)],
   ...: [0, 1, 0],
   ...: [-sin(theta), 0, cos(theta)],
   ...: ])

In [6]: rot_x = Matrix([
   ...: [1, 0, 0],
   ...: [0, cos(theta), -sin(theta)],
   ...: [0, sin(theta), cos(theta)],
   ...: ])

In [7]: R = rot_z.subs(theta, gamma) * rot_y.subs(theta, beta) * rot_x.subs(theta, alpha)

In [8]: alpha_val = np.radians(5)

In [9]: beta_val = np.radians(10)

In [10]: gamma_val = np.radians(15)

In [11]: mat_symbolic_truth = R.subs(alpha, alpha_val).subs(beta, beta_val).subs(gamma, gamma_val)

In [12]: from scipy.spatial.transform import Rotation

In [13]: mat = Rotation.from_euler("ZYX", [gamma_val, beta_val, alpha_val]).as_matrix()

In [14]: np.allclose(mat, np.array(mat_symbolic_truth, dtype=float))
Out[14]: True

In [15]: mat_extrinsic = Rotation.from_euler("xyz", [alpha_val, beta_val, gamma_val]).as_matrix()

In [16]: np.allclose(mat_extrinsic, np.array(mat_symbolic_truth, dtype=float))
Out[16]: True
```

Other permutations don't work:

```python
In [17]: mat_err1 = Rotation.from_euler("XYZ", [alpha_val, beta_val, gamma_val]).as_matrix()

In [18]: np.allclose(mat_err1, np.array(mat_symbolic_truth, dtype=float))
Out[18]: False

In [19]: mat_err2 = Rotation.from_euler("XYZ", [gamma_val, beta_val, alpha_val]).as_matrix()

In [20]: np.allclose(mat_err2, np.array(mat_symbolic_truth, dtype=float))
Out[20]: False

In [21]: mat_err3 = Rotation.from_euler("ZYX", [alpha_val, beta_val, gamma_val]).as_matrix()

In [22]: np.allclose(mat_err3, np.array(mat_symbolic_truth, dtype=float))
Out[22]: False

In [23]: mat_extrinsic_err1 = Rotation.from_euler("zyx", [alpha_val, beta_val, gamma_val]).as_matrix()

In [24]: np.allclose(mat_extrinsic_err1, np.array(mat_symbolic_truth, dtype=float))
Out[24]: False

In [25]: mat_extrinsic_err2 = Rotation.from_euler("zyx", [gamma_val, beta_val, alpha_val]).as_matrix()

In [26]: np.allclose(mat_extrinsic_err2, np.array(mat_symbolic_truth, dtype=float))
Out[26]: False

In [27]: mat_extrinsic_err3 = Rotation.from_euler("xyz", [gamma_val, beta_val, alpha_val]).as_matrix()

In [28]: np.allclose(mat_extrinsic_err3, np.array(mat_symbolic_truth, dtype=float))
Out[28]: False
```

To be certain, I checked the implementation which is found in
https://github.com/scipy/scipy/blob/main/scipy/spatial/transform/_rotation.pyx,
where the `Rotation.from_euler` method is implemented in terms of quaternion composition
using `_elementary_quat_compose`, the definition of which is:

```python
cdef double[:, :] _elementary_quat_compose(
    const uchar[:] seq, const double[:, :] angles, bint intrinsic=False
):
    cdef double[:, :] result = _make_elementary_quat(seq[0], angles[:, 0])
    cdef Py_ssize_t seq_len = seq.shape[0]

    for idx in range(1, seq_len):
        if intrinsic:
            result = _compose_quat(
                result,
                _make_elementary_quat(seq[idx], angles[:, idx]))
        else:
            result = _compose_quat(
                _make_elementary_quat(seq[idx], angles[:, idx]),
                result)
    return result
```

So, intrinsic rotations are composed in forward order, extrinsic rotations in reverse order.
I was also confused by this at first, so the documentation could certainly clarify this better.

[^1]: The reasons:
    - In aeronautics, the angles are usually ordered by the right-handed roll-pitch-yaw system,
      so the order of α, β, γ would be reversed.
    - The first and second matrices, both denoted by R in that section,
      are not equivalent according to the definitions.
    - The last paragraph is unclear and the way I read it only applies to the second R,
      for the first R, the "order of operations" is left-to-right.

**Comment by j-bowhay** [comment]
Friendly ping @scottshambaugh do you have a view here?

**Comment by scottshambaugh** [comment]
(I don't use Euler angles day-to-day so had to dig in here) - I believe @adigitoleo is right in that the "order of operations" is flipped for extrinsic and intrinsic rotations. To add one more example to the pile:

```python
from scipy.spatial.transform import Rotation as R

phi = 30; theta = 20; omega = 45
rx = R.from_euler("xyz", [0, 0, omega], degrees=True)
ry = R.from_euler("xyz", [0, theta, 0], degrees=True)
rz = R.from_euler("xyz", [phi, 0, 0], degrees=True)

r3 = R.from_euler("xyz", [phi, theta, omega], degrees=True)
print('extrinsic xyz')
print((rx.as_matrix() @ ry.as_matrix() @ rz.as_matrix()))  # note the order
print(r3.as_matrix())

r3 = R.from_euler("XYZ", [phi, theta, omega], degrees=True)
print('intrinsic XYZ')
print((rz.as_matrix() @ ry.as_matrix() @ rx.as_matrix()))  # note the order
print(r3.as_matrix())
```

```
extrinsic xyz
[[ 0.66446302 -0.49145005  0.5629971 ]
 [ 0.66446302  0.73329482 -0.14410968]
 [-0.34202014  0.46984631  0.81379768]]
[[ 0.66446302 -0.49145005  0.5629971 ]
 [ 0.66446302  0.73329482 -0.14410968]
 [-0.34202014  0.46984631  0.81379768]]

intrinsic XYZ
[[ 0.66446302 -0.66446302  0.34202014]
 [ 0.73329482  0.49145005 -0.46984631]
 [ 0.14410968  0.5629971   0.81379768]]
[[ 0.66446302 -0.66446302  0.34202014]
 [ 0.73329482  0.49145005 -0.46984631]
 [ 0.14410968  0.5629971   0.81379768]]
```

The docs are definitely not clear here and could use some explanation. That explanation should also include *why* this is set up the way it is, with an example use case that shows the utility of the inversion. It's too late to change the convention and break backwards compatibility.

@RichardLee0211 your results are due to this inversion of the multiplication order. Your plot should read $r_2 = R^z_\omega R^y_\theta$, and then the results make sense.

**Comment by scottshambaugh** [comment]
(I don't use Euler angles day-to-day so had to dig in here) - I believe @adigitoleo is right in that the "order of operations" is flipped for extrinsic and intrinsic rotations. To add one more example to the pile:

```python
from scipy.spatial.transform import Rotation as R

phi = 30; theta = 20; omega = 45
rx = rX = R.from_euler("xyz", [phi, 0, 0], degrees=True)
ry = rY = R.from_euler("xyz", [0, theta, 0], degrees=True)
rz = rZ = R.from_euler("xyz", [0, 0, omega], degrees=True)

r3 = R.from_euler("xyz", [phi, theta, omega], degrees=True)
print('extrinsic xyz')
# note the order - rx is applied first to a vector, then ry, then rz
print((rz.as_matrix() @ ry.as_matrix() @ rx.as_matrix()))
print(r3.as_matrix(), '\n')

r3 = R.from_euler("XYZ", [phi, theta, omega], degrees=True)
print('intrinsic XYZ')
# note the order - rZ is applied first to a vector, then rY, then rX
print((rX.as_matrix() @ rY.as_matrix() @ rZ.as_matrix()))
print(r3.as_matrix())
```

```
extrinsic xyz
[[ 0.66446302 -0.49145005  0.5629971 ]
 [ 0.66446302  0.73329482 -0.14410968]
 [-0.34202014  0.46984631  0.81379768]]
[[ 0.66446302 -0.49145005  0.5629971 ]
 [ 0.66446302  0.73329482 -0.14410968]
 [-0.34202014  0.46984631  0.81379768]] 

intrinsic XYZ
[[ 0.66446302 -0.66446302  0.34202014]
 [ 0.73329482  0.49145005 -0.46984631]
 [ 0.14410968  0.5629971   0.81379768]]
[[ 0.66446302 -0.66446302  0.34202014]
 [ 0.73329482  0.49145005 -0.46984631]
 [ 0.14410968  0.5629971   0.81379768]]
 ```

The docs are definitely not clear here and could use some explanation. That explanation should also include *why* this is set up the way it is, with an example use case that shows the utility of the inversion. It's too late to change the convention and break backwards compatibility.

@RichardLee0211 your results are as expected, and due to this inversion of the multiplication order. Your plot should read $r_2 = R^z_\omega R^y_\theta$, and then the results make sense.

**Comment by cfriedland5** [comment]
I was just looking at this and my interpretation is that extrinsic/intrinsic aren't mixed up but the sequence is applied from right to left mirroring how rotation matrices are applied. It wasn't what I was expecting but makes sense in a way—it should be documented better at any rate.

rotation = Rotation.from_euler("XYZ", [90, 90, 0], degrees=True)
vector = [1, 0, 0]
rotation.apply(vector)
# [0., 1., 0.]

rotation_1 = Rotation.from_euler("Z", 0, degrees=True)
rotation_2 = Rotation.from_euler("Y", 90, degrees=True)
rotation_3 = Rotation.from_euler("X", 90, degrees=True)

(rotation_3 * rotation_2 * rotation_1).apply(vector)
# Also [0., 1., 0.]
rotation_3.apply(rotation_2.apply(rotation_1.apply(vector)))
# Also [0., 1., 0.] as well

**Comment by cfriedland5** [comment]
@adbugger @larsoner @nmayorov @rgommers it looks like you worked on #8945 which introduced Rotation.from_euler(). I know it's been awhile but do you have any insight about what the order of *seq* is supposed to mean and how the documentation for intrinsic vs extrinsic rotations should be fixed?

**Comment by adbugger** [comment]
Looking over the code from my original implementation, I think I can clarify a few points here:

- `Rotation.as_matrix` returns the final rotation matrix which must be pre-multiplied to a vector to perform the rotation
- `Rotation.from_euler(seq, angles)` describes a sequence of rotations around the given axes with the given angles:
  - `Rotation.from_euler("xyz", [a, b, c])` represents a series of extrinsic rotations, first around x-axis with angle a, then around y-axis with angle b, and finally around z-axis with angle c. Since the rotations are all extrinsic, the frame of reference (x,y,z axes) remains constant throughout.
  - `Rotation.from_euler("XYZ", [a,b,c])` represents a series of _intrinsic_ rotations. First rotation is around x-axis with angle a. Second rotation is around y'-axis with angle b. This y'-axis is body centered and is actually the y-axis (from external coordinate frame) with the first rotation applied. The final third rotation is around z''-axis with angle c. This z''-axis is again body-centered and is actually the z-axis (from external coordinate frame) with the first and second rotations applied.

If `R_x`, `R_y`, `R_z` represent _extrinsic_ rotations around x, y, z with specific angles, then `Rotation.from_euler("xyz", angles)` is the result of applying `R_x`, `R_y`, `R_z` in sequence, i.e. `R_1 = R_z * R_y * R_x`.

Similarly `Rotation.from_euler("XYZ", angles)` is the result of applying 3 consecutive _intrinsic_ rotations `R_X`, `R_Y`, `R_Z` in sequence, where the X,Y,Z axes are in the body centred frame. The resultant rotation matrix `R_2 = R_Z * R_Y * R_X`. Please note that since the axes are changing `R_X`, `R_Y` and `R_Z` matrices are not equivalent in general to `R_x`, `R_y`, `R_z` matrices.

In both cases "xyz" and "XYZ" the sequence of rotations is the same. First along x-axis, then y-axis, then z-axis.

I think the source of the "order reversal" confusion happens because mathematically `R_2` is equivalent to `R_1` with the order switched. `R_Z * R_Y * R_X = R_x * R_y * R_z`. The full rotation (sequence of intrinsic rotations) is equivalent to the simplified external rotations with the sequence switched.

@evbernardes has a better and more beautiful explanation of the math here https://github.com/scikit-image/scikit-image/pull/5263#issuecomment-1422357031

Looking at the gist:

> A gist with a different check is provided here: https://gist.github.com/whophil/c809692ceb8d5b4a22b70edc103dd87c, but arrives at the same conclusion.

@whophil I think a comparison between your gist and the linked issue comment will be illuminating. Specifically in the gist, the same matrices cannot be used for both intrinsic and extrinsic rotations. For extrinsic rotations the coordinate axes are not changing, but for intrinsic rotations the coordinate axes do change so to express the final coordinates in the extrinsic frame after an _intrinsic_ rotation requires a different rotation matrix. It also explains why the rotation composition in the gist is applicable for external but not internal rotations.

Apologies for the long comment. I do believe evbernardes' linked comment clarifies all of this in a much neater fashion.

Finally, the relationship between internal and external rotations is also tested here:
https://github.com/scipy/scipy/blob/d55cb95282b5cfab381e373d3c127630c7e915a0/scipy/spatial/transform/tests/test_rotation.py#L558-L565

In summary:

1. The rotations are always in the order specified by `seq`. The confusion arises from the fact that while the resultant of intrinsic rotations is the same as extrinsic with order switched, the individual rotations in each are not the same.
2. I am not sure what documentation changes are exactly required here as I think the docs are already correct. However an explanation similar to https://github.com/scikit-image/scikit-image/pull/5263#issuecomment-1422357031 would clear things up quite a bit.

**Comment by evbernardes** [comment]
For what it's worth, I also banged my head against the wall until I did the math myself, as I couldn't find a good explanation on this "order reversal" (which is really **not** a reversal in convention, just a mathematical quirk).

So yeah, +1 for maybe adding an explanation to the docs.

**Comment by cfriedland5** [comment]
Thanks for that explanation.

I think I've figured what was tripping me up and some ideas for improving the documentation. I was working with a coordinate transforming/passive rotation but `Rotation` represents an active rotation, so I needed to invert the rotation to make it agree with my textbook:

```python
rotation_i_was_expecting = Rotation.from_euler("ZYX", [yaw, pitch, roll]).inv()
```

`Rotation` representing an active rotation is only mentioned in from_quat() and as_quat() (and kind of in apply()) and I think it would be helpful to briefly mention that in every from/as method. I also think it would help to add a paragraph in the class-level documentation about how for a rotation that transforms coordinates (passive) rather than rotates actual position (active) the `Rotation` object needs to be inverted or apply()'s *inverse* parameter needs to be true.
